File: client/typescript/src/controller.ts
import { ApiError, SandboxConfig, SandboxRef } from "./schemas";
import { Sandbox, SandboxError, toError } from "./sandbox";
import { parseSSE } from "./sse";

/** Gateway URL used when none is supplied and `HIVER_GATEWAY_URL` is unset. */
export const DEFAULT_GATEWAY_URL = "http://localhost:10000";

/** Env var that overrides {@link DEFAULT_GATEWAY_URL} when no explicit URL is given. */
export const GATEWAY_URL_ENV = "HIVER_GATEWAY_URL";

/**
 * Resolve the gateway base URL. An explicit `gatewayUrl` (e.g. from
 * {@link GatewayOptions}) always wins; otherwise the `HIVER_GATEWAY_URL` env var
 * is used, falling back to {@link DEFAULT_GATEWAY_URL}. The env lookup is guarded
 * so the client also works where `process` is undefined (e.g. the browser).
 */
export function resolveGatewayUrl(gatewayUrl?: string): string {
  const fromEnv =
    typeof process !== "undefined" ? process.env?.[GATEWAY_URL_ENV] : undefined;
  return gatewayUrl ?? fromEnv ?? DEFAULT_GATEWAY_URL;
}

const SANDBOX_KEY_PATTERN = /^[A-Za-z0-9_-]{1,64}$/;

/**
 * Logical image name the gateway routes on when `config.image` is unset. The
 * gateway matches the `x-hiver-image` header against its per-image clusters, so
 * every create must carry one; this is the fallback.
 */
export const DEFAULT_IMAGE_NAME = "agent-base";


export interface GatewayOptions {
  /**
   * Base URL of the gateway. When omitted, the `HIVER_GATEWAY_URL` env var is
   * used, else {@link DEFAULT_GATEWAY_URL} (`http://localhost:10000`).
   */
  gatewayUrl?: string;
  /** Override the global fetch (e.g. for testing or custom transports). */
  fetch?: typeof fetch;
  /**
   * Timeout in milliseconds for controller operations. Defaults to 60s. Pass
   * `0` to disable timeouts.
   */
  timeoutMs?: number;
}

const DEFAULT_TIMEOUT_MS = 60_000;

/**
 * Create a sandbox, or fetch the existing one when `key` is already in use.
 * The key acts as an idempotency key: calling again with the same key returns
 * the same sandbox and leaves the supplied `config` unapplied. Resolves once
 * the sandbox is ready to accept requests.
 */
/**
 * Apply the client-side config defaults every create carries: an unset `fs`
 * becomes the standard `/workspace` local mount and an unset `egress` opens all
 * hosts. Shared by {@link getOrCreateSandbox} and `allowSandbox` — a config
 * pinned into an egress override replaces the nested create's body verbatim, so
 * it must carry the same defaults a direct create would, or the sandbox comes
 * up without a workspace while a resumed VM snapshot still holds the 9p mount.
 */
export function sandboxConfigWithDefaults(
  config: SandboxConfig = {},
): SandboxConfig {
  // Drop keys whose value is null/undefined so an explicitly-null field (e.g.
  // `fs: null` from a serialized config where an unset optional became null)
  // falls back to the default below instead of clobbering it and failing
  // validation (`.optional()` accepts undefined, not null).
  const provided = Object.fromEntries(
    Object.entries(config).filter(([, v]) => v != null),
  );
  return SandboxConfig.parse({
    fs: [
      {
        backend: "local",
        mount: "/workspace",
        acls: [{ path: "/workspace/**", access: "rw" }],
      },
    ],
    egress: [{ host: "*", access: "allow" }],
    ...provided,
  });
}

export async function getOrCreateSandbox(
  key: string,
  config: SandboxConfig = {},
  opts: GatewayOptions = {},
): Promise<Sandbox> {
  if (!SANDBOX_KEY_PATTERN.test(key)) {
    throw new Error(
      `getOrCreateSandbox: key ${JSON.stringify(key)} must match ${SANDBOX_KEY_PATTERN}`,
    );
  }
  const validated = sandboxConfigWithDefaults(config);
  const gatewayBase = resolveGatewayUrl(opts.gatewayUrl).replace(/\/+$/, "");
  const fetchImpl = opts.fetch ?? fetch;
  const timeout = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;

  const imageName = config.image ?? DEFAULT_IMAGE_NAME;
  const base = gatewayBase;

  try {
    return await provisionSandbox(
      key,
      validated,
      base,
      imageName,
      fetchImpl,
      timeout,
    );
  } catch (err) {
    if (err instanceof SandboxError && err.status === 0 && timeout > 0) {
      return provisionSandbox(
        key,
        validated,
        base,
        imageName,
        fetchImpl,
        timeout,
      );
    }
    throw err;
  }
}

async function provisionSandbox(
  key: string,
  config: SandboxConfig,
  base: string,
  imageName: string,
  fetchImpl: typeof fetch,
  timeout: number,
): Promise<Sandbox> {
  let res: Response;
  try {
    res = await fetchImpl(
      `${base}/v1/sandboxes/${encodeURIComponent(key)}`,
      {
        method: "POST",
        headers: {
          "content-type": "application/json",
          "x-hiver-image": imageName,
          // The gateway consistent-hashes the create onto a pack host by this
          // header (the image clusters' MAGLEV hash_policy), so every
          // get-or-create for a key lands on the same pod. The key is also in
          // the path, but Envoy hashes on the header.
          "x-hiver-key": key,
        },
        body: JSON.stringify(config),
        signal: timeout > 0 ? AbortSignal.timeout(timeout) : undefined,
      },
    );
  } catch (err) {
    if (isConnectionRefused(err)) {
      throw new SandboxError(
        "getOrCreateSandbox",
        0,
        `gateway is not reachable at ${base} (connection refused). Is it running?`,
      );
    }
    throw err;
  }
  if (res.status !== 200 && res.status !== 201) {
    const text = await res.text();
    let body: { error: string; details?: Record<string, unknown> } | undefined;
    try {
      body = ApiError.parse(JSON.parse(text));
    } catch {
      body = undefined;
    }
    throw new SandboxError(
      "getOrCreateSandbox",
      res.status,
      body?.error ?? text ?? res.statusText,
      body,
    );
  }
  const ref = SandboxRef.parse(await res.json());
  return new Sandbox(ref, { gatewayUrl: base, fetch: fetchImpl });
}

/**
 * List all currently running sandboxes.
 */
export async function listSandboxes(
  opts: GatewayOptions = {},
): Promise<Sandbox[]> {
  const base = resolveGatewayUrl(opts.gatewayUrl).replace(/\/+$/, "");
  const fetchImpl = opts.fetch ?? fetch;
  const timeout = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;

  let res: Response;
  try {
    res = await fetchImpl(`${base}/controller/v1/sandboxes`, {
      signal: timeout > 0 ? AbortSignal.timeout(timeout) : undefined,
    });
  } catch (err) {
    if (isConnectionRefused(err)) {
      throw new SandboxError(
        "listSandboxes",
        0,
        `gateway is not reachable at ${base} (connection refused). Is it running?`,
      );
    }
    throw err;
  }
  if (res.status !== 200) {
    throw await toError(res, "listSandboxes");
  }
  // Tolerate a null/absent body (e.g. an empty list serialized as JSON null).
  const body = ((await res.json()) as unknown[] | null) ?? [];
  const refs = body.map((r) => SandboxRef.parse(r));
  return refs.map(
    (ref) => new Sandbox(ref, { gatewayUrl: base, fetch: fetchImpl }),
  );
}

/** Lifecycle transition a sandbox can go through. */
export type SandboxLifecycleStatus = "start" | "stop" | "die" | "destroy";
/** A lifecycle change observed for a single sandbox. */
export interface SandboxLifecycleEvent {
  /** Server-assigned unique identifier (uuid). */
  id: string;
  /** Caller-chosen key the sandbox was provisioned under. */
  key: string;
  /** Which lifecycle transition occurred. */
  status: SandboxLifecycleStatus;
}

/**
 * Watch lifecycle changes (start, stop, die, destroy) across all sandboxes as
 * they happen. Yields events until `signal` is aborted or the stream ends.
 */
export async function* watchSandboxEvents(
  opts: GatewayOptions = {},
  signal?: AbortSignal,
): AsyncGenerator<SandboxLifecycleEvent, void, void> {
  const base = resolveGatewayUrl(opts.gatewayUrl).replace(/\/+$/, "");
  const fetchImpl = opts.fetch ?? fetch;

  let res: Response;
  try {
    res = await fetchImpl(`${base}/controller/v1/sandboxes/events`, {
      headers: { Accept: "text/event-stream" },
      signal,
    });
  } catch (err) {
    if (isConnectionRefused(err)) {
      throw new SandboxError(
        "watchSandboxEvents",
        0,
        `gateway is not reachable at ${base} (connection refused). Is it running?`,
      );
    }
    throw err;
  }

  if (!res.ok || !res.body) {
    throw new SandboxError("watchSandboxEvents", res.status, res.statusText);
  }

  for await (const frame of parseSSE(res.body, signal)) {
    try {
      yield JSON.parse(frame.data) as SandboxLifecycleEvent;
    } catch {
      // skip malformed frames
    }
  }
}

function isConnectionRefused(err: unknown): boolean {
  if (!(err instanceof Error)) return false;
  const cause = (err as { cause?: unknown }).cause;
  if (hasCode(cause, "ECONNREFUSED")) return true;
  if (cause instanceof AggregateError) {
    return cause.errors.some((e) => hasCode(e, "ECONNREFUSED"));
  }
  return false;
}

function hasCode(e: unknown, code: string): boolean {
  return (
    typeof e === "object" &&
    e !== null &&
    (e as { code?: unknown }).code === code
  );
}

File: client/go/controller.go
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"
)

// DefaultGatewayURL is the URL used when no gateway URL is provided.
const DefaultGatewayURL = "http://localhost:10000"

var sandboxKeyPattern = regexp.MustCompile(`^[A-Za-z0-9_-]{1,64}$`)

const defaultTimeout = 30 * time.Second

// Client is the controller client. It provisions and manages sandboxes
// through the Hive gateway.
type Client struct {
	baseURL string
	http    *http.Client
	timeout time.Duration
}

// Option configures a Client.
type Option func(*Client)

// WithHTTPClient overrides the underlying HTTP client.
func WithHTTPClient(hc *http.Client) Option {
	return func(c *Client) { c.http = hc }
}

// WithTimeout sets the default per-operation timeout applied to point
// requests (non-streaming). Pass 0 to disable. Defaults to 30s.
func WithTimeout(d time.Duration) Option {
	return func(c *Client) { c.timeout = d }
}

// NewClient creates a controller client pointed at gatewayURL.
func NewClient(gatewayURL string, opts ...Option) *Client {
	c := &Client{
		baseURL: strings.TrimRight(gatewayURL, "/"),
		http:    &http.Client{},
		timeout: defaultTimeout,
	}
	for _, o := range opts {
		o(c)
	}
	return c
}

// GetOrCreateSandbox creates a sandbox, or fetches the existing one when key
// is already in use. The key acts as an idempotency key: calling again with
// the same key returns the same sandbox and leaves config unapplied. It blocks
// until the sandbox is ready (up to the client's configured timeout).
//
// As a convenience, an empty config.FS defaults to a /workspace local mount
// with rw access, and an empty config.Egress opens egress to all hosts.
func (c *Client) GetOrCreateSandbox(ctx context.Context, key string, config SandboxConfig) (*Sandbox, error) {
	if !sandboxKeyPattern.MatchString(key) {
		return nil, fmt.Errorf("GetOrCreateSandbox: key %q must match %s", key, sandboxKeyPattern)
	}

	if len(config.FS) == 0 {
		config.FS = []FileSystem{
			{Mount: "/workspace", Backend: "local", ACLs: []ACLRule{{Path: "/workspace/**", Access: "rw"}}},
		}
	}
	if len(config.Egress) == 0 {
		config.Egress = []EgressRule{{Access: "allow", Host: "*"}}
	}

	body, err := json.Marshal(config)
	if err != nil {
		return nil, fmt.Errorf("GetOrCreateSandbox: encode config: %w", err)
	}

	ctx, cancel := c.withTimeout(ctx)
	defer cancel()

	url := fmt.Sprintf("%s/controller/v1/sandboxes/%s", c.baseURL, key)
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("GetOrCreateSandbox: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return nil, fmt.Errorf("GetOrCreateSandbox: %w", readAPIError(resp))
	}

	var ref SandboxRef
	if err := json.NewDecoder(resp.Body).Decode(&ref); err != nil {
		return nil, fmt.Errorf("GetOrCreateSandbox: decode response: %w", err)
	}

	sbx := newSandbox(ref, c.baseURL, c.http)

	if c.timeout > 0 {
		if err := waitUntilReachable(ctx, sbx, c.timeout); err != nil {
			return nil, err
		}
	}

	return sbx, nil
}

// ListSandboxes returns all currently provisioned sandboxes.
func (c *Client) ListSandboxes(ctx context.Context) ([]*Sandbox, error) {
	ctx, cancel := c.withTimeout(ctx)
	defer cancel()

	url := fmt.Sprintf("%s/controller/v1/sandboxes", c.baseURL)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ListSandboxes: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("ListSandboxes: %w", readAPIError(resp))
	}

	var refs []SandboxRef
	if err := json.NewDecoder(resp.Body).Decode(&refs); err != nil {
		return nil, fmt.Errorf("ListSandboxes: decode response: %w", err)
	}

	sandboxes := make([]*Sandbox, len(refs))
	for i, ref := range refs {
		sandboxes[i] = newSandbox(ref, c.baseURL, c.http)
	}
	return sandboxes, nil
}

// Shutdown permanently stops and removes the sandbox identified by key. It is
// idempotent for already-stopped sandboxes; unknown keys return an error.
func (c *Client) Shutdown(ctx context.Context, key string) error {
	ctx, cancel := c.withTimeout(ctx)
	defer cancel()

	url := fmt.Sprintf("%s/controller/v1/shutdown/%s", c.baseURL, key)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, nil)
	if err != nil {
		return err
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("Shutdown: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNoContent {
		return nil
	}
	return fmt.Errorf("Shutdown: %w", readAPIError(resp))
}

// WatchEvents streams sandbox lifecycle events (start/stop/die/destroy) from
// the controller. The returned channel is closed when ctx is done or the
// server closes the stream. Any final error is sent on errc.
func (c *Client) WatchEvents(ctx context.Context) (<-chan SandboxLifecycleEvent, <-chan error) {
	events := make(chan SandboxLifecycleEvent)
	errc := make(chan error, 1)

	go func() {
		defer close(events)

		url := fmt.Sprintf("%s/controller/v1/sandboxes/events", c.baseURL)
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
		if err != nil {
			errc <- err
			return
		}
		req.Header.Set("Accept", "text/event-stream")

		resp, err := c.http.Do(req)
		if err != nil {
			errc <- fmt.Errorf("WatchEvents: %w", err)
			return
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			errc <- fmt.Errorf("WatchEvents: %w", readAPIError(resp))
			return
		}

		for frame := range readSSE(resp.Body) {
			var evt SandboxLifecycleEvent
			if err := json.Unmarshal([]byte(frame.data), &evt); err != nil {
				continue
			}
			select {
			case events <- evt:
			case <-ctx.Done():
				return
			}
		}
	}()

	return events, errc
}

// withTimeout returns a child context that inherits the parent's deadline
// unless it already has one, in which case the parent is returned unchanged.
func (c *Client) withTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	if c.timeout == 0 {
		return ctx, func() {}
	}
	if _, ok := ctx.Deadline(); ok {
		return ctx, func() {}
	}
	return context.WithTimeout(ctx, c.timeout)
}

func readAPIError(resp *http.Response) error {
	body, _ := io.ReadAll(resp.Body)
	var apiErr APIError
	if json.Unmarshal(body, &apiErr) == nil && apiErr.Message != "" {
		return &apiErr
	}
	return fmt.Errorf("status %d: %s", resp.StatusCode, strings.TrimSpace(string(body)))
}

func waitUntilReachable(ctx context.Context, s *Sandbox, timeout time.Duration) error {
	deadline := time.Now().Add(timeout)
	var lastErr error
	for time.Now().Before(deadline) {
		pingCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
		err := s.Ping(pingCtx)
		cancel()
		if err == nil {
			return nil
		}
		lastErr = err
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(200 * time.Millisecond):
		}
	}
	return fmt.Errorf("GetOrCreateSandbox: sandbox %s did not become reachable within %s: %w",
		s.ID, timeout, lastErr)
}
